### PR TITLE
 Build Realtek r8125 driver version 9.010.01 for XCP

### PR DESCRIPTION
**Pull request by andrew64k**
Signed-off-by: Andrew Lindh <andrew@netplex.net>

### CODE REVIEW

File: SOURCES/0001-config_change.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+-ENABLE_PTP_SUPPORT = n
+-ENABLE_PTP_MASTER_MODE = n
+-ENABLE_RSS_SUPPORT = n
++ENABLE_MULTIPLE_TX_QUEUE = y
++ENABLE_PTP_SUPPORT = y
++ENABLE_PTP_MASTER_MODE = y
++ENABLE_RSS_SUPPORT = y
+ ENABLE_LIB_SUPPORT = n
+-ENABLE_USE_FIRMWARE_FILE = n
++ENABLE_USE_FIRMWARE_FILE = y
```
Comment (stormi) [inline SOURCES/0001-config_change.patch:27]: Why is this change needed? Why do these features need to be enabled? (Please explain like I don't know this hardware... Which is true :)).

File: SOURCES/0002-use_new_api.patch [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
++++ b/src/r8125.h	2022-05-06 02:36:55.000000000 +0000
+@@ -51,11 +51,9 @@
+ typedef int netdev_tx_t;
+ #endif
+ 
+-/*
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)&& !defined(ENABLE_LIB_SUPPORT)
+ #define RTL_USE_NEW_INTR_API
+ #endif
+-*/
```
Comment (stormi) [inline SOURCES/0002-use_new_api.patch:15]: What does this change, and why is it needed?

File: SOURCES/0003-sg-tso.patch [hunk 3]
```diff
@@ -0,0 +1,18 @@
+--- a/src/r8125_n.c	2022-06-28 15:32:07.000000000 +0000
++++ b/src/r8125_n.c	2022-07-31 19:25:19.317518176 +0000
+@@ -12703,6 +12703,7 @@
+                 tp->cp_cmd |= RxChkSum;
+ #else
+                 dev->features |= NETIF_F_RXCSUM;
++                dev->features |= NETIF_F_SG | NETIF_F_TSO;
```
Comment (stormi) [inline SOURCES/0003-sg-tso.patch]: Same question as the previous ones.

File: SOURCES/0002-use_new_api.patch (hunk 2 above)
Comment (stormi) [reply SOURCES/0002-use_new_api.patch:15]: How does commenting out the whole 3 lines enable `RTL_USE_NEW_INTR_API`? As I read it, the condition is already true: kernel version is >= 4.12 and ENABLE_LIB_SUPPORT is not defined.

Comment (andrew64k) [reply SOURCES/0002-use_new_api.patch:15]: Because it's uncommenting the lines and enabling them.

Comment (stormi) [reply SOURCES/0002-use_new_api.patch:15]: Oh, right, those are removed comments, not added ones. Any idea why they were commented out in the first place?

**Comment by andrew64k** [comment]
Realtek 8125 firmware is available in "linux-firmware" on kernel.org

**Comment by stormi** [comment]
So we also need to package the firmware?

**Comment by andrew64k** [comment]
Optional, not required but recommended. I don't see it in the standard XCP CentOS linux-firware package.

The 8125 interfaces have a base firmware as part of the chip. So they should work without a firmware blob.
As the firmware is part of current linux-firmware packages it would seem wise to include it for XCP.

It could just be added to this package (driver and firmware). It could be added to the XCP linux-firmware package (like others).
It could be added to a realtek-firmware package (with other realtek firmware too). It could be downloaded automatically (wget) from kernal.org.

**Comment by stormi** [comment]
This PR looks good to me. Is there anything you want to change before I merge it?

Regarding firmware, XCP-ng 8.3's `linux-firmware` RPM will already contain:
```
/lib/firmware/rtl_nic/rtl8125a-3.fw
/lib/firmware/rtl_nic/rtl8125b-1.fw
/lib/firmware/rtl_nic/rtl8125b-2.fw
```

On XCP-ng 8.2 they're missing, but we could add them to the `linux-firmware` RPM, as well as other missing firmware if that's really useful. As long as we're just adding firmware, we can add it to the `linux-firmware` package. If we start *updating* firmware, then that's a different story, as there always is a risk of regression. Which doesn't mean we can't do it, but it will have to be pondered case by case.

**Comment by andrew64k** [comment]
They have not updated the firmware in a while. I'm using it for testing. If they update it, then it can wait for a normal firmware update cycle. It would be useful to have the official Linux Realtek firmware on XCP so things are consistent.

I have a new 2.5G switch and new Intel/AMD machines in my test lab so I'll run a new full Host/Guest/NFS/iSCSI/VLAN/MTU/usage/performance test on the driver.

**Comment by andrew64k** [comment]
I'm going to revert some of the code additions. Their driver does not support everything with 4.19 and it still has some performance issues in testing. It works, but it's not great...

**Comment by stormi** [comment]
If it's good enough for installing and people can update afterwards, maybe we should merge this first PR and include it in the public alpha ISOs for XCP-ng?

**Comment by andrew64k** [comment]
I'll remove some of the changes tonight so it's ready for alpha inclusion.

Same issue/question as the I225 (igc) driver as to how it needs to be included on the installer ISO.

**Comment by stormi** [comment]
Maybe tonight? Else, if it's the last step remaining towards building alpha ISOs, I might merge as is tomorrow.

**Comment by andrew64k** [comment]
New version updated. I removed the code change that enables SG and TSO as TSO does not work correctly with the current code and kernel 4.19 and is recommended to be off by Realtek. The driver still has some performance issues but it works on my test machines.

**Review by stormi — APPROVED** [review]

**Comment by stormi** [comment]
Thanks @andrew64k!

Koji build: http://koji.xcp-ng.org/taskinfo?taskID=26491
